internal/component/prometheus/remote/queue/network/loop.go

```diff
@@ -0,0 +1,374 @@
+}
+
+func (l *loop) sendingCleanup() {
+	types.PutTimeSeriesSliceIntoPool(l.series)
+	l.sendBuffer = l.sendBuffer[:0]
+	l.series = make([]*types.TimeSeriesBinary, 0, l.batchCount)
+	l.lastSend = time.Now()
+}
+
+// send is the main work loop of the loop.
+func (l *loop) send(ctx context.Context, retryCount int) sendResult {
+	result := sendResult{}
+	defer func() {
+		recordStats(l.series, l.isMeta, l.statsFunc, result, len(l.sendBuffer))
+	}()
+	var err error
+	// Check to see if this is a retry and we can reuse the buffer.
+	// I wonder if we should do this, its possible we are sending things that have exceeded the TTL.
+	if len(l.sendBuffer) == 0 {
+		var data []byte
+		var wrErr error
+		if l.isMeta {
+			data, wrErr = createWriteRequestMetadata(l.req, l.series, l.buf)
+		} else {
+			data, wrErr = createWriteRequest(l.req, l.series, l.externalLabels, l.buf)
+		}
+		if wrErr != nil {
+			result.err = wrErr
+			result.recoverableError = false
+			return result
+		}
+		l.sendBuffer = snappy.Encode(l.sendBuffer, data)
+	}
+
+	httpReq, err := http.NewRequest("POST", l.cfg.URL, bytes.NewReader(l.sendBuffer))
+	if err != nil {
+		result.err = err
+		result.recoverableError = true
+		result.networkError = true
+		return result
+	}
+	httpReq.Header.Add("Content-Encoding", "snappy")
+	httpReq.Header.Set("Content-Type", "application/x-protobuf")
+	httpReq.Header.Set("User-Agent", l.cfg.UserAgent)
+	httpReq.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+	httpReq.SetBasicAuth(l.cfg.Username, l.cfg.Password)
+
+	if retryCount > 0 {
+		httpReq.Header.Set("Retry-Attempt", strconv.Itoa(retryCount))
+	}
+	ctx, cncl := context.WithTimeout(ctx, l.cfg.Timeout)
+	defer cncl()
+	resp, err := l.client.Do(httpReq.WithContext(ctx))
+	// Network errors are recoverable.
+	if err != nil {
+		result.err = err
+		result.networkError = true
+		result.recoverableError = true
+		result.retryAfter = l.cfg.RetryBackoff
+		return result
+	}
+	result.statusCode = resp.StatusCode
+	defer resp.Body.Close()
+	// 500 errors are considered recoverable.
+	if resp.StatusCode/100 == 5 || resp.StatusCode == http.StatusTooManyRequests {
+		result.retryAfter = retryAfterDuration(l.cfg.RetryBackoff, resp.Header.Get("Retry-After"))
+		result.recoverableError = true
+		return result
+	}
+	// Status Codes that are not 500 or 200 are not recoverable and dropped.
+	if resp.StatusCode/100 != 2 {
+		scanner := bufio.NewScanner(io.LimitReader(resp.Body, 1_000))
+		line := ""
+		if scanner.Scan() {
+			line = scanner.Text()
+		}
+		result.err = fmt.Errorf("server returned HTTP status %s: %s", resp.Status, line)
+		return result
+	}
+
+	result.successful = true
+	return result
+}
+
+func createWriteRequest(wr *prompb.WriteRequest, series []*types.TimeSeriesBinary, externalLabels map[string]string, data *proto.Buffer) ([]byte, error) {
+	if cap(wr.Timeseries) < len(series) {
+		wr.Timeseries = make([]prompb.TimeSeries, len(series))
+	}
+	wr.Timeseries = wr.Timeseries[:len(series)]
```

Comment:
The `wr.Timeseries` lives throughout the loop lifetime, but the `series` is returned to the pool and re-allocated in `sendingCleanUp` - what's the idea here? why?

Reply:
TimeSeries is a protobuf concept and only exist within the lifecycle of the loop and are reused. Whereas series are used by other parts of the full remote write and component.

Reply:
TimeSeries is a loaded term, TimeSerieskbinary might be to close to the protobuf name so will consider if there is a better name.